### ISSUE TITLE
Types don't work with different versions of hardhat (v2.3.0 and v2.5.0)

### ISSUE THREAD

**Issue by jjperezaguinaga** [paulrberg/hardhat-packager#2]
![image](https://user-images.githubusercontent.com/1128312/128023932-334b3385-5d14-4263-9e44-e6a51f13985d.png)

For some reason after installing the plugin I see the following error in my `hardhat.config.ts`. You can see changes in this PR - austintgriffith/scaffold-eth/pull/300

**Comment by paulrberg** [comment]
Yes, there was an issue with the types in v1.0.0 and v1.0.1. But all further versions should be fine. Can you double-check that you're using v1.0.4?

Also, check that you imported the plugin at the top of the `hardhat.config.ts` file. See [how](https://github.com/hifi-finance/hifi-protocol/blob/67120aecc9d264b66ce9261c255207b8b13bb61f/hardhat.config.ts#L3) I do it in hifi-protocol.

**Comment by StErMi** [comment]
Hi @paulrberg I've installed the latest version of the package, added it as the top of the import and I still get the same problem as @jjperezaguinaga. I've currently installed the `"hardhat-packager": "^1.0.4"` version

**Comment by StErMi** [comment]
Also, if the import order is important you should specify that in the readme so people will not create more issue about that.

**Comment by StErMi** [comment]
The complete error is this:

```
Type '{ solidity: string; packager: { contracts: string[]; }; paths: { artifacts: string; cache: string; sources: string; tests: string; }; networks: { hardhat: { chainId: number; }; }; typechain: { outDir: string; target: string; }; }' is not assignable to type 'HardhatUserConfig'.
  Object literal may only specify known properties, and 'packager' does not exist in type 'HardhatUserConfig'.ts(2322)
  ```

**Comment by paulrberg** [comment]
Thanks guys for the detailed report. I identified the issue, will ship a fix soon.

**Comment by paulrberg** [comment]
**Update**: false positive. I didn't identify the issue, in that I can't reproduce it. Sorry if this is a hassle for you, but could any one of you reproduce the type error in a GitHub repo that I can clone?

Thank you.

**Comment by paulrberg** [comment]
The `HardhatUserConfig` is correctly overridden here:

https://github.com/paulrberg/hardhat-packager/blob/main/src/type-extensions.ts

Which is part of the distributed files on npm:

https://unpkg.com/browse/hardhat-packager@1.0.4/dist/index.d.ts

As per the `types` field in `package.json`:

https://unpkg.com/browse/hardhat-packager@1.0.4/package.json

**Comment by jjperezaguinaga** [comment]
@paulrberg Thanks for the reply, let me see if I can get you more information to replicate the issue. FWIW I had a branch using the JavaScript version that didn't show the error

**Comment by jjperezaguinaga** [comment]
Update here, I'm indeed using version `1.0.4`, here's the `shashum1` of the `dist` folder after you install it:

```
gitpod /workspace/scaffold-eth $ sha1sum packages/hardhat/node_modules/hardhat-packager/dist/*  | awk '{ print $1; }' | sha1sum 
0991856b2b166f5cc7f9956a031f28324c58e669  -
```

which it's the same as the one from the HiFi protocol after you run `yarn`

```
gitpod /workspace/hifi-protocol $ sha1sum node_modules/hardhat-packager/dist/*  | awk '{ print $1; }' | sha1sum 
0991856b2b166f5cc7f9956a031f28324c58e669  -
```

However, I `ls`'ed the contents of the folder (which happens to use `yarn workspaces`) and noticed a `node_modules` folder:

```
gitpod /workspace/scaffold-eth $ ls -halt packages/hardhat/node_modules/hardhat-packager/
total 52K
drwxr-xr-x   8 gitpod gitpod 4.0K Aug  4 13:33 node_modules
drwxr-xr-x 761 gitpod gitpod  28K Aug  4 13:33 ..
drwxr-xr-x   2 gitpod gitpod 4.0K Aug  4 13:33 dist
drwxr-xr-x   4 gitpod gitpod 4.0K Aug  4 13:33 .
-rw-r--r--   1 gitpod gitpod  150 Aug  4 13:32 LICENSE.md
-rw-r--r--   1 gitpod gitpod 2.8K Aug  4 13:32 package.json
-rw-r--r--   1 gitpod gitpod 3.2K Aug  4 13:32 README.md
gitpod /workspace/scaffold-eth $ ls -halt packages/hardhat/node_modules/hardhat-packager/node_modules/
total 32K
drwxr-xr-x  2 gitpod gitpod 4.0K Aug  4 13:33 .bin
drwxr-xr-x  8 gitpod gitpod 4.0K Aug  4 13:33 .
drwxr-xr-x  9 gitpod gitpod 4.0K Aug  4 13:33 hardhat
drwxr-xr-x  5 gitpod gitpod 4.0K Aug  4 13:33 js-sha3
drwxr-xr-x  2 gitpod gitpod 4.0K Aug  4 13:33 ms
drwxr-xr-x  3 gitpod gitpod 4.0K Aug  4 13:33 bn.js
drwxr-xr-x  4 gitpod gitpod 4.0K Aug  4 13:33 ..
drwxr-xr-x 20 gitpod gitpod 4.0K Aug  4 13:32 @ethersproject
```

which in contrast the hife project doesn't.

```
gitpod /workspace/hifi-protocol $ ls -halt node_modules/hardhat-packager/
total 60K
drwxr-xr-x 1050 gitpod gitpod  36K Aug  4 13:42 ..
drwxr-xr-x    3 gitpod gitpod 4.0K Aug  4 13:41 .
drwxr-xr-x    2 gitpod gitpod 4.0K Aug  4 13:41 dist
-rw-r--r--    1 gitpod gitpod  150 Aug  4 13:41 LICENSE.md
-rw-r--r--    1 gitpod gitpod 2.8K Aug  4 13:41 package.json
-rw-r--r--    1 gitpod gitpod 3.2K Aug  4 13:41 README.md
```

I deleted the `node_modules` folder in my repo and compared the `sha1sum` until it was equal to hifi:

```
gitpod /workspace/hifi-protocol $ find node_modules/hardhat-packager/ -type f -print0 | sort -z | xargs -0 sha1sum  | awk '{print $1}'
9f5084efbda8f88fcdfc6bca30dbc09462d91c36
dd27664d9e02781b85583e3bf404982e3336434e
27f14a291009d5fa9131eb2f482f202c716fa1b1
cb4adf3a19ca13383b6ef7e847a1ae87b7cf23d0
21039d2130d55a18bf7fe57b638a0861dcf77753
c13aecf36b135d6e4941dc2ea8bc72d7744b7f43
ce1cd5a5628e439df3885ea3bf0d9b1bec6637ba
41fda93ac2d04a23754ccb0a0795fbbe90b421e4
aa278d8296ced096f50ef33355390feb5b7ea79a
486a267f5531b104ed8f46395cb786526c5ea563
9a1a6beeacfb3fde5be624a5b146a4fb88ab02a5
1c78c67d4013a94b1ed0fde317d8d9be1eda3683
5c680b9d4af99f7fe4fa1393533e249a639daae0
9bd7157c50b49594e4b5669f65408a6d03fce91f
4ea83cb63ae8bb9372c8c71d71bc82002da5030d
f4c2b3b7d8ec75121abf757a09f751e32f71e2c4
30842aff6e18b7ea247b785b199eff0cd70b557c
b724e4f966b0f2bf051c7d593dc900f017a1753d
397eb4a51c892fdf6722316e2ac727798c672638
gitpod /workspace/hifi-protocol $ find node_modules/hardhat-packager/ -type f -print0 | sort -z | xargs -0 sha1sum  | awk '{print $1}'| sha1sum 
1c7f49be77fa6b721740466715f11f5554c1fb4d  -
```

here's my own project file contents
```
gitpod /workspace/scaffold-eth $ find packages/hardhat/node_modules/hardhat-packager/ -type f -print0 | sort -z | xargs -0 sha1sum  | awk '{print $1}'9f5084efbda8f88fcdfc6bca30dbc09462d91c36
dd27664d9e02781b85583e3bf404982e3336434e
27f14a291009d5fa9131eb2f482f202c716fa1b1
cb4adf3a19ca13383b6ef7e847a1ae87b7cf23d0
21039d2130d55a18bf7fe57b638a0861dcf77753
c13aecf36b135d6e4941dc2ea8bc72d7744b7f43
ce1cd5a5628e439df3885ea3bf0d9b1bec6637ba
41fda93ac2d04a23754ccb0a0795fbbe90b421e4
aa278d8296ced096f50ef33355390feb5b7ea79a
486a267f5531b104ed8f46395cb786526c5ea563
9a1a6beeacfb3fde5be624a5b146a4fb88ab02a5
1c78c67d4013a94b1ed0fde317d8d9be1eda3683
5c680b9d4af99f7fe4fa1393533e249a639daae0
9bd7157c50b49594e4b5669f65408a6d03fce91f
4ea83cb63ae8bb9372c8c71d71bc82002da5030d
f4c2b3b7d8ec75121abf757a09f751e32f71e2c4
30842aff6e18b7ea247b785b199eff0cd70b557c
b724e4f966b0f2bf051c7d593dc900f017a1753d
397eb4a51c892fdf6722316e2ac727798c672638
gitpod /workspace/scaffold-eth $ find packages/hardhat/node_modules/hardhat-packager/ -type f -print0 | sort -z | xargs -0 sha1sum  | awk '{print $1}'| sha1sum 
1c7f49be77fa6b721740466715f11f5554c1fb4d  -
```

Afterwards, it worked properly:

```
gitpod /workspace/scaffold-eth $ yarn package
yarn run v1.22.10
$ yarn workspace nextjs-ethereum-starter-hardhat package
$ npx hardhat prepare-package
Downloading compiler 0.8.3
Compiling 3 files with 0.8.3
Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.
--> contracts/Multicall.sol


Generating typings for: 3 artifacts in dir: ../frontend/types/typechain for target: ethers-v5
Successfully generated 7 typings!
Compilation finished successfully
Preparing 1 contracts ...
Successfully prepared 1 contracts for registry deployment!
Done in 7.82s.
gitpod /workspace/scaffold-eth $ 
```

tl;dr: A project configuration might throw an error. Check hoisting properties to avoid this.

As soon as I find a workaround in my branch, will share the next update. For now, a good recommendation is to remove any `node_modules` inside the `hardhat-packager` folder.

**Comment by StErMi** [comment]
I'm not using yarn workspaces. I'll try to do a node_modules cleanup and reinstall from scratch.

**Comment by StErMi** [comment]
Nope, even after a fresh install (I'm using npm and I've removed node_modules and package-lock.json) I still get the same error.

This is my current dev dep list

```
  "devDependencies": {
    "@nomiclabs/hardhat-ethers": "^2.0.2",
    "@nomiclabs/hardhat-solhint": "^2.0.0",
    "@nomiclabs/hardhat-waffle": "^2.0.1",
    "@openzeppelin/contracts": "^4.1.0",
    "@typechain/ethers-v5": "^7.0.1",
    "@typechain/hardhat": "^2.2.0",
    "@types/chai": "^4.2.19",
    "@types/mocha": "^8.2.2",
    "@types/node": "^15.12.4",
    "chai": "^4.3.4",
    "ethereum-waffle": "^3.4.0",
    "ethers": "^5.3.1",
    "hardhat": "2.4.1",
    "hardhat-packager": "^1.0.4",
    "mocha": "^7.2.0",
    "prettier": "^2.3.1",
    "prettier-plugin-solidity": "^1.0.0-beta.13",
    "solhint": "^2.3.1",
    "solhint-plugin-prettier": "^0.0.5",
    "solidity-coverage": "^0.7.16",
    "ts-node": "^10.0.0",
    "typechain": "^5.1.2",
    "typescript": "^4.3.4"
  }
  ```

**Comment by paulrberg** [comment]
I figured it out. You guys are using an older version of `hardhat`. @jjperezaguinaga is using v2.3.0, while @StErMi is using v2.4.1.

I upgraded to v2.3.0 on @jjperezaguinaga's fork of `scaffold-eth`, and the type error went away. This must be an inconsistency between the different versions of `hardhat` and their associated `HardhatUserConfig` type definitions.

**Comment by paulrberg** [comment]
Just shipped [v1.0.5](https://github.com/paulrberg/hardhat-packager/releases/tag/v1.0.5) which bumps the version of the `hardhat` peer dependency to `^2.5.0`.

**Comment by StErMi** [comment]
I can confirm it's fixed now ;)

**Comment by paulrberg** [comment]
Great! Thanks guys.

**Comment by jjperezaguinaga** [comment]
Confirmed issue and fix, thanks!

**Comment by sshmaxime** [comment]
I have this issue with `1.1.0`

**Comment by paulrberg** [comment]
Hi @sshmaxime, GitHub was down for like ~1h today. Can you recheck now?

If you're still encountering the type error, can you share what version of Hardhat you are using? You can find that out by running `yarn why hardhat`.

**Comment by sshmaxime** [comment]
Ok it works with hh `2.6.0`, I was using hh `2.5.0`

**Comment by paulrberg** [comment]
It should have worked with `2.5.0` as well.. Are you sure you were on that version? can you still reproduce the type error?

**Comment by sshmaxime** [comment]
I kinda moved on .. i'll try again later if I get some free time

**Comment by paulrberg** [comment]
Ok. If it continues to be a problem others should report it too.
